src/main/java/org/broadinstitute/hellbender/utils/iterators/SamReaderQueryingIterator.java

```diff
@@ -0,0 +1,159 @@
+        this.intervalQueryPerformed = false;
+        this.unmappedQueryPerformed = false;
+
+        this.currentIterator = loadNextIterator();
+        this.nextRecord = loadNextRecord();
+    }
+
+    /**
+     * Converts a List of SimpleIntervals into the format required by the SamReader query API
+     * @param rawIntervals SimpleIntervals to be converted
+     * @return A sorted, merged list of QueryIntervals suitable for passing to the SamReader query API
+     */
+    private QueryInterval[] prepareQueryIntervals( final List<SimpleInterval> rawIntervals ) {
+        if ( rawIntervals == null || rawIntervals.isEmpty() ) {
+            return null;
+        }
+
+        // This might take a while with large interval lists, so log a status message
+        logger.info("Preparing intervals for traversal");
+
+        // Convert each SimpleInterval to a QueryInterval
+        QueryInterval[] convertedIntervals = new QueryInterval[rawIntervals.size()];
+        int intervalIndex = 0;
+        for ( final SimpleInterval rawInterval : rawIntervals ) {
+            convertedIntervals[intervalIndex] = IntervalUtils.convertSimpleIntervalToQueryInterval(rawInterval, reader.getFileHeader().getSequenceDictionary());
+            ++intervalIndex;
+        }
+
+        // Intervals must be optimized (sorted and merged) in order to use the htsjdk query API
+        return QueryInterval.optimizeIntervals(convertedIntervals);
+    }
+
+    private SAMRecord loadNextRecord() {
+        if ( currentIterator == null ) {
+            return null;
+        }
+
+        while ( currentIterator != null && ! currentIterator.hasNext() ) {
+            currentIterator = loadNextIterator();
+        }
+
+        return currentIterator != null ? currentIterator.next() : null;
+    }
+
+    private CloseableIterator<SAMRecord> loadNextIterator() {
+        // The SamReader API requires us to close out the previous iterator over our reader before opening a new one.
+        if ( currentIterator != null ) {
+            currentIterator.close();
+        }
+
+        if ( hasQueryIntervals() && ! intervalQueryPerformed ) {
+            intervalQueryPerformed = true;
+            return reader.queryOverlapping(queryIntervals);
+        }
+        else if ( queryUnmapped && ! unmappedQueryPerformed ) {
+            unmappedQueryPerformed = true;
+            return reader.queryUnmapped();
+        }
+
+        return null;
+    }
+
+    private boolean hasQueryIntervals() {
+        return queryIntervals != null && queryIntervals.length > 0;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextRecord != null;
+    }
+
+    @Override
+    public SAMRecord next() {
+        if ( nextRecord == null ) {
+            throw new NoSuchElementException("Iterator is exhausted");
+        }
+
+        final SAMRecord toReturn = nextRecord;
+        nextRecord = loadNextRecord();
+        return toReturn;
+    }
+
+    @Override
+    public void close() {
```

Comment:
I'm not sure we care -- we mainly want to ensure that we have the ability to close out open iterators when we need to, which the existing implementation enables.